### PR TITLE
Add natural transformation synonym & operator

### LINKED ISSUE #23: Testing
This applies to all the core libraries, not just prelude, but this the first one in the hierarchy where the problem of circular `devDependencies` occurs.

I've been thinking about ways we might resolve this, one approach we had previously was to just remove the offending duplicate from `bower_components` when building, and actually, something along these lines is by far my preference, although this time using a `postinstall` script in `.bowerrc` so it happens automatically as part of the install process.

It's so much simpler than any other alternative, with basically no downsides I can see (other than being aesthetically offensive :wink:) - `bower` certainly doesn't care if you have circular dependencies.

If you `bower install x` something where `x` has `devDependencies` the `devDependencies` are not installed. `bower install` in the current project, without the `--production` flag is the only time `devDependencies` are brought in, and in this situation `.bowerrc` will be picked up for the project, so we can definitely do the removal there.

All the other ways I've looked at, having a separate `bower.json` in `tests/`, etc. make it really awkward to run tests as part of the normal `gulp` or `grunt` build (actually, impossible in the `grunt` case, as you have to run `grunt` in the directory the `Gruntfile.js` lives) and then you end up having to maintain two sets of bower dependencies too.

### PR DESCRIPTION
**Pull request by garyb**
Resolves #23 

Once this is merged `purescript-free` will need an update, as currently that library provides a module `Data.NaturalTransformation`.

### CODE REVIEW

File: src/Data/NaturalTransformation.purs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+-- |
+-- | A natural transformation allows a functor to be transformed into another
+-- | without changing the internal structure. This is a little similar to the
+-- | `map` operation of a `Functor`, although "one level up". In the case of
+-- | `map`, the functor is preserved while the inner values are re-mapped,
+-- | whereas here the inner values are preserved and the functor is changed
+-- | instead.
+type NaturalTransformation f g = forall a. f a -> g a
+
+infixr 4 type NaturalTransformation as ~>
```
Comment (garyb) [inline src/Data/NaturalTransformation.purs:19]: I figured this gives us enough room to slide `->` underneath while still not making it the lowest precedence, when we decide what that should be.

File: src/Data/NaturalTransformation.purs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module Data.NaturalTransformation where
+
+-- | A type for natural transformations between functors (although the `Functor`
```
Comment (paf31) [inline src/Data/NaturalTransformation.purs]: I wouldn't mention `Functor` here, and just say "between type constructors" instead. `runCoyoneda` doesn't need this to be a `Functor`, for example.

File: src/Data/NaturalTransformation.purs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+-- | constraint is not enforced).
+-- |
+-- | A natural transformation allows a functor to be transformed into another
```
Comment (paf31) [inline src/Data/NaturalTransformation.purs]: I'm not sure about the connection with `map`. I'd say something like "transforms a functor" and something about uniformity, and give an example like `listToMaybe`, which I think provides a decent intuition.

File: src/Data/NaturalTransformation.purs (hunk 1 above)
Comment (paf31) [inline src/Data/NaturalTransformation.purs:19]: 👍 Sounds good.

**Comment by garyb** [comment]
Updated the description. I used `fromFoldable` as an example as we don't have `listToMaybe`.

I do think it's worth mentioning that our ~> does not require functor instances though, since technically it should if we're going to call it a natural transformation. I moved it to a note at the end rather than being part of the up front explanation instead though.

**Comment by hdgarrood** [comment]
Does parametricity already ensure that, for a natural transformation `f`, we have `f >>> map g == map g >>> f`?

**Comment by texastoland** [comment]
Is it not enforced because you might want a `NaturalTransformation` without that constraint or something else?

**Comment by paf31** [comment]
👍 Looks good to me, thanks!

**Comment by garyb** [comment]
@AppShipIt yeah, there are lots of things where the constraint would be unhelpful, Phil's coyoneda example being one. 

`Coyoneda f` gives you a `Functor` for any `f`, regardless of whether it has a `Functor` instance or not, so it would not be possible to write the type of `lowerCoyenda` as `Coyoneda f ~> f` without imposing `Functor f`, which kinda defeats the point.

I could be wrong here, so please someone jump in if my ignorance is showing, but I think any type of kind `* -> *` can be considered a lowercase-f-functor regardless of whether there is a `Functor` instance, because our `Functor` is more restrictive than that, as it's for endofunctors.

**Comment by garyb** [comment]
Hmm, maybe not, as this is the natural transformation of endofunctors too, the category is fixed as `(->)`.

I don't know :wink:

**Comment by texastoland** [comment]
Thanks @garyb 🙏

**Pull request merged**